### ISSUE TITLE
Making thenOpen sync

### ISSUE THREAD

**Issue by robin-obrien** [n1k0/casperjs#1474]
Let's say I have a function in my code that contains the step thenOpen which determines the return value. How can I make the function wait for the thenOpen to complete and only then return the value?
What is the best workaround for this scenario?

**Comment by robin-obrien** [comment]
Please find two examples: First, with refactoring code in functions and the second as a long function (In real life, it quickly becomes spaghetti code)

Example ONE

```
var casper = require('casper').create({
    verbose: true,
    logLevel: "debug"
});

function importantProcessing(){
    this.log("Entering importantProcessing");
    this.thenOpen("http://phantomjs.org/", function processPhantomjs(){
        return "Processed Data";
    });
    this.log("Exiting importantProcessing");
}

function insertToDB(processedData){
    this.log("Inserting the data "+processedData);
}

casper.start("http://casperjs.org/", function startStep(){
    var processedData = null;
    this.then(function importantProcessingStep() {
        processedData = importantProcessing();
    });
    this.then(function insertToDBStep() {
        insertToDB(processedData);
    });
});

casper.run();

```
Prints the following in console:
[info] [phantom] Starting...
[info] [phantom] Running suite: 2 steps
[debug] [phantom] opening url: http://casperjs.org/, HTTP GET
[debug] [phantom] Navigation requested: url=http://casperjs.org/, type=Other, willNavigate=true, isMainFrame=true
[debug] [phantom] url changed to "http://casperjs.org/"
[debug] [phantom] Successfully injected Casper client-side utilities
[info] [phantom] Step startStep 2/2 http://casperjs.org/ (HTTP 200)
[info] [phantom] Step startStep 2/2: done in 1835ms.
[info] [phantom] Step importantProcessingStep 3/4 http://casperjs.org/ (HTTP 200)
[debug] [phantom] Navigation requested: url=about:blank, type=Other, willNavigate=true, isMainFrame=true
[debug] [phantom] url changed to "about:blank"


Example TWO
```
var casper = require('casper').create({
    verbose: true,
    logLevel: "debug"
});

casper.start("http://casperjs.org/", function startStep(){
    var processedData = null;
    this.then(function importantProcessingStep() {
        this.log("Entering importantProcessing");
        this.thenOpen("http://phantomjs.org/", function processPhantomjs(){
            processedData = "Processed Data";
        });
        this.log("Exiting importantProcessing");
    });

    this.then(function insertToDBStep() {
        this.log("Inserting the data "+processedData);
    });
});

casper.run();

```
Prints the following in console:
[info] [phantom] Starting...
[info] [phantom] Running suite: 2 steps
[debug] [phantom] opening url: http://casperjs.org/, HTTP GET
[debug] [phantom] Navigation requested: url=http://casperjs.org/, type=Other, willNavigate=true, isMainFrame=true
[debug] [phantom] url changed to "http://casperjs.org/"
[debug] [phantom] Successfully injected Casper client-side utilities
[info] [phantom] Step startStep 2/2 http://casperjs.org/ (HTTP 200)
[info] [phantom] Step startStep 2/2: done in 2523ms.
[info] [phantom] Step importantProcessingStep 3/4 http://casperjs.org/ (HTTP 200)
[debug] [phantom] Entering importantProcessing
[debug] [phantom] Exiting importantProcessing
[info] [phantom] Step importantProcessingStep 3/4: done in 2543ms.
[debug] [phantom] opening url: http://phantomjs.org/, HTTP GET
[debug] [phantom] Navigation requested: url=http://phantomjs.org/, type=Other, willNavigate=true, isMainFrame=true
[debug] [phantom] url changed to "http://phantomjs.org/"
[debug] [phantom] Successfully injected Casper client-side utilities
[info] [phantom] Step processPhantomjs 5/6 http://phantomjs.org/ (HTTP 200)
[info] [phantom] Step processPhantomjs 5/6: done in 3647ms.
[info] [phantom] Step insertToDBStep 6/6 http://phantomjs.org/ (HTTP 200)
[debug] [phantom] Inserting the data Processed Data
[info] [phantom] Step insertToDBStep 6/6: done in 3670ms.
[info] [phantom] Done 6 steps in 3686ms
[debug] [phantom] Navigation requested: url=about:blank, type=Other, willNavigate=true, isMainFrame=true
[debug] [phantom] url changed to "about:blank"

**Comment by robin-obrien** [comment]
@n1k0 @istr   Can anyone help

**Comment by robin-obrien** [comment]
Anyone still maintaining the code or this repo is dead?

**Comment by mepard** [comment]
It was really hard for me to get my head around when code runs in casperjs and how to get things to happen when I want, particularly how to get results out. It looks like you've figured it out, though, with Example Two, it's just that you don't like the solution.

The problem is that the casperjs/phantomjs/javascript stack is inherently asynchronous. Each function runs in a callback in response to an event or setTimeout expiring. You can't do synchronous operations because you must return from the current function to get future functions to run. Look at http://docs.casperjs.org/en/latest/faq.html#how-does-then-and-the-step-stack-work, which links to http://blog.thepete.net/blog/2011/07/02/javascript-promises/, which links to http://www.infoq.com/articles/surviving-asynchronous-programming-in-javascript.

And you can see for yourself that the repo is not dead by checking for recent activity. Maybe it's not as much activity as we'd like, but so it goes.

**Comment by istr** [comment]
@mepard Thanks for pointing that out:

> It was really hard for me to get my head around when code runs in casperjs and how to get things to happen when I want

[...]

> The problem is that the casperjs/phantomjs/javascript stack is inherently asynchronous.

@robin-obrien Considering all the issues you filed against CasperJS up to now, all qualifiy as "help request". From what I read it seems that you still don't accept that javascript (and testing it) is in fact asynchronous, like @mepard pointed out.

Even though I had very long and very routine professional experience with other programming languages before starting to seriously use JavaScript, it took some time for me too, to wrap my head around the implications of this fact.

It would be nice if you could address your questions that are mostly non-issues to [the mailing list](https://groups.google.com/forum/#!forum/casperjs) as per https://github.com/n1k0/casperjs/blob/master/README.md

This project is currently maintained without any sponsorship in the spare time of a handful of developers. Please respect this and help us to focus on real issues, that need to be addressed.

Closing this as a duplicate of #1432. Thanks for your understanding.
